File: backend/apps/cart/management/commands/release_expired_reservations.py
import logging

from django.core.management.base import BaseCommand

from apps.cart.services import release_expired_reservations

logger = logging.getLogger(__name__)


class Command(BaseCommand):
    help = 'Release expired cart stock reservations.'

    def handle(self, *args, **options):
        released = release_expired_reservations()
        self.stdout.write(self.style.SUCCESS(f'Released {released} expired cart reservation(s).'))

File: app/Console/Commands/ReleaseExpiredReservations.php
<?php

namespace App\Console\Commands;

use App\Models\Cart;
use App\Services\CartService;
use Illuminate\Console\Command;

class ReleaseExpiredReservations extends Command
{
    protected $signature = 'cart:release-expired-reservations';

    protected $description = 'Release product stock reserved by abandoned checkout sessions.';

    public function handle(CartService $cartService): int
    {
        $count = 0;

        Cart::query()
            ->whereNotNull('reserved_until')
            ->where('reserved_until', '<=', now())
            ->with('items.product')
            ->chunkById(100, function ($carts) use ($cartService, &$count): void {
                foreach ($carts as $cart) {
                    $cartService->releaseStock($cart);
                    $count++;
                }
            });

        $this->info("Released {$count} expired cart reservation(s).");

        return self::SUCCESS;
    }
}
